### PR TITLE
[#2851] feat(core): Add relational backend for Group Entity

### PR DESCRIPTION
**Pull request by lw-yang**
### What changes were proposed in this pull request?

add relational backend for Group Entity

[note]: wait for #2850 merged first

### Why are the changes needed?

Fix: #2851 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

ut

### CODE REVIEW

File: core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java [hunk 1]
```diff
@@ -748,4 +779,82 @@ public static RolePO initializeRolePOWithVersion(RoleEntity roleEntity, RolePO.B
...
+  /**
+   * Update GroupPO version
+   *
+   * @param oldGroupPO the old GroupPO object
+   * @param newGroup the new GroupEntity object
+   * @return GroupPO object with updated version
+   */
+  public static GroupPO updateGroupPOWithVersion(GroupPO oldGroupPO, GroupEntity newGroup) {
+    Long lastVersion = oldGroupPO.getLastVersion();
+    // Will set the version to the last version + 1 when having some fields need be multiple version
```
Comment (yuqi1129) [inline core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:814]: This comment is confusing, I know that version value is useless here, but as you describe, you will increment it in the comment, but you won't do it accordingly in the code below.

File: core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java [hunk 2]
```diff
@@ -692,6 +695,34 @@ public static UserEntity fromUserPO(UserPO userPO, List<RolePO> rolePOs, Namespa
...
+  public static GroupEntity fromGroupPO(
+      GroupPO groupPO, List<RolePO> rolePOs, Namespace namespace) {
+    try {
+      List<String> roleNames =
+          rolePOs.stream().map(RolePO::getRoleName).collect(Collectors.toList());
+      List<Long> roleIds = rolePOs.stream().map(RolePO::getRoleId).collect(Collectors.toList());
+      return GroupEntity.builder()
+          .withId(groupPO.getGroupId())
+          .withName(groupPO.getGroupName())
+          .withRoleNames(roleNames.isEmpty() ? null : roleNames)
```
[the remark this answers is not in the document]
Comment (lw-yang) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:715]: returns null in kv store, keep consistent with the behavior of kv

File: core/src/main/java/com/datastrato/gravitino/storage/relational/service/UserMetaService.java [hunk 3]
```diff
@@ -193,34 +193,41 @@ public <E extends Entity & HasIdentifier> UserEntity updateUser(
...
-          if (deleteRoleIds.isEmpty()) {
-            return;
-          }
-          SessionUtils.doWithoutCommit(
-              UserRoleRelMapper.class,
-              mapper ->
-                  mapper.softDeleteUserRoleRelByUserAndRoles(
-                      newEntity.id(), Lists.newArrayList(deleteRoleIds)));
-        });
+                    mapper.batchInsertUserRoleRel(
```
[the remark this answers is not in the document]
Comment (lw-yang) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/service/UserMetaService.java:212]: it means user-role relation 
https://github.com/datastrato/gravitino/blob/9429d1bbb23bcb86d66d0bf360f4050f38d85971/scripts/mysql/schema-0.5.0-mysql.sql#L148

In fact, i just added exception handing here

File: core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java (hunk 1 above)
Comment (lw-yang) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:814]: the version field is reserved and is not currently used, it same as other entity

File: core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuqi1129) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:715]: @qqqttt123 
Can you explain the reason here?

[the remark this answers is not in the document]
Comment (qqqttt123) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:715]: You would better refer to 
```
Builder builder =  GroupEntity.builder()
          .withId(groupPO.getGroupId())
          .withName(groupPO.getGroupName());
if (!roleNames.isEmpty()) {
    builder.withRoleNames(roleNames);
}
builder.build();
```

File: core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java (hunk 1 above)
Comment (qqqttt123) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/utils/POConverters.java:814]: You should use
```
// TODO: set the version to the last version + 1 when having some fields need be multiple version
```

**Review by qqqttt123 — APPROVED** [review]
LGTM.

**Comment by jerryshao** [comment]
@yuqi1129 please take another look at this PR.

File: core/src/main/java/com/datastrato/gravitino/storage/relational/service/GroupMetaService.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+              && GroupEntity.namespace().levels().length == 3,
+          "The identifier should not be null and should have three level.");
+
+      Long metalakeId =
+          MetalakeMetaService.getInstance().getMetalakeIdByName(GroupEntity.namespace().level(0));
+      GroupPO.Builder builder = GroupPO.builder().withMetalakeId(metalakeId);
+      GroupPO GroupPO = POConverters.initializeGroupPOWithVersion(GroupEntity, builder);
+
+      List<Long> roleIds = GroupEntity.roleIds();
+      if (roleIds == null) {
```
Comment (yuqi1129) [inline core/src/main/java/com/datastrato/gravitino/storage/relational/service/GroupMetaService.java:98]: As `roleNames` and `rolesIds` are identical in null or not null,  so I wander if we need the logic here.

Comment (lw-yang) [reply core/src/main/java/com/datastrato/gravitino/storage/relational/service/GroupMetaService.java:98]: ok, i will optimize it

**Comment by yuqi1129** [comment]
@qqqttt123 
I have no further comments, please go ahead if you are okay with the changes.

**Review by qqqttt123 — APPROVED** [review]

**Comment by qqqttt123** [comment]
Thanks all. Wait for CI.

**Comment by jerryshao** [comment]
We'd better modify to support storing role, user, and group under and {metalake}.system.{role|user|group}.

**Comment by jerryshao** [comment]
> @jerryshao I agree with you, it would also be aligned with other entities' relation storage.
> 
> I am working on this, but i found the code is changed a lot, could i modify this after 0.5.0 release

OK, no problem.

**Comment by lw-yang** [comment]
sure,  I have created an issue #3099

**Comment by qqqttt123** [comment]
Merged this first. Thanks all.

**Pull request merged**